### ISSUE TITLE
[2.1] Segmentation Model Definition (U-Net / DeepLabV3+)

### ISSUE THREAD

**Issue by poltergeistjoya** [poltergeistjoya/Vacant_Lot_Detection#12]
## Summary
Build U-Net and DeepLabV3+ models via `segmentation-models-pytorch` library. Support configurable encoder + pretrained weights.

## Implementation Details
- **New file:** `vacant_lot/segmentation.py`

### Model Builder
```python
def build_model(arch, in_channels=8, encoder_name='resnet34', encoder_weights='imagenet', classes=1):
    '''Build U-Net or DeepLabV3+ with pretrained encoder.
    - arch: 'unet' or 'deeplabv3plus'
    - in_channels: 8 (RGBN + 4 indices)
    - encoder_weights: 'imagenet' or None
    - classes: 1 (binary vacant/non-vacant)
    '''
```

### Loss Function: Combined BCE + Dice
```python
class SegmentationLoss(nn.Module):
    def __init__(self, pos_weight=10, bce_weight=0.5, dice_weight=0.5):
        '''
        - BCE: Per-pixel loss, upweights positive (vacant) class
        - Dice: Region-based metric, inherently handles imbalance
        - Ignore mask: label=255 → zero loss contribution
        '''
```

**Rationale:**
- BCE provides stable per-pixel gradients
- Dice directly optimizes overlap metric
- Together handle imbalance better than either alone
- For Dice: exclude label=255 from both numerator AND denominator

### Metrics
```python
def compute_metrics(preds, targets, threshold=0.5):
    return {
        'iou': IoU,           # Primary metric
        'dice': Dice,         # F1-like
        'precision': TP/(TP+FP),
        'recall': TP/(TP+FN),
        'kappa': Cohen's Kappa,
    }
```

### Early Stopping
```python
class EarlyStopping:
    def __init__(self, patience=10, mode='max'):
        # Monitor val IoU; stop if no improvement for 10 epochs
```

## Architecture Choice: U-Net vs DeepLabV3+
- **U-Net** (Ronneberger et al., 2015): Standard, simple, works well
- **DeepLabV3+** (Chen et al., 2018): Multi-scale ASPP, typically 2–5% higher mIoU
- **Plan:** Start with U-Net; run DeepLabV3+ as one-parameter comparison

## Pretrained Encoders Justification
ImageNet features in early layers (edges, textures) transfer broadly to remote sensing. Fine-tuning adapts later layers while retaining low-level representations (Nogueira et al., 2017; Yosinski et al., 2014).

## Acceptance Criteria
- [ ] U-Net model builds without errors
- [ ] DeepLabV3+ model builds without errors
- [ ] Loss function computes correctly with ignore mask (label=255)
- [ ] Metrics compute correctly on dummy tensors
- [ ] Early stopping works: training stops when val metric plateaus
- [ ] Architecture summaries printed (parameter counts)

## Dependencies
- segmentation-models-pytorch >= 0.3.3
- torch >= 2.0
- torchvision >= 0.15

## References
- Ronneberger et al. (2015), U-Net, MICCAI, arXiv:1505.04597
- Chen et al. (2018), DeepLabV3+, ECCV, arXiv:1802.02611
- Iakubovskii (2019), segmentation_models.pytorch on GitHub
- Jadon (2020), A survey of loss functions for semantic segmentation
- Yeung et al. (2022), Unified Focal loss
